{% extends 'base.html' %}
{% block title %}Pipeline{% endblock %}
{% block content %}
<link rel="stylesheet" href="../static/pipeline.css">

    <div class="pipeline">
        
        <div class="step">
            <h3>1. Check DNA or RNA Quality with FastQC</h3>
            <p>Run FastQC on two FastQ files to evaluate the DNA or RNA quality. FastQC generates an HTML file showcasing the DNA quality.</p>
        </div>
        <div class="step">
            <h3>2. Improve DNA Quality with Trimmomatic</h3>
            <p>If DNA quality is poor, use Trimmomatic to remove adapters and enhance DNA quality. Run FastQC again to confirm improvements.</p>
        </div>
        <div class="step">
            <h3>3. Genome Mapping with BWA</h3>
            <p>Use BWA to create a genome map (index) with "hg19.fa." Utilize BWA mem to map FastQ files against the reference genome, generating a .sam file with mapping information.</p>
        </div>
        <div class="step">
            <h3>4. Data Sorting and Cleaning with Picard</h3>
            <p>Employ Picard's functions (SortSam, FixMateInformation, AddOrReplaceReadGroups) for data organization and error reduction. Use MarkDuplicates to remove duplicates, PrintReads to filter low-quality reads, ensuring data integrity.</p>
        </div>
        <div class="step">
            <h3>5. Genetic Variant Calling with GATK</h3>
            <p>Use GATK's HaplotypeCaller tool to identify genomic variations. IndexFeatureFile and VariantFiltration refine variations, and the results are stored in VCF files.</p>
        </div>
        <div class="step">
            <h3>6. ANNOVAR for Genomic Analysis</h3>
            <p>Convert filtered VCF files to AVinput format using annovar_convert. Utilize ANNOVAR to annotate genomic variations with the humandb database.</p>
        </div>
        <div class="step">
            <h3>7. Final Steps</h3>
            <p>Run IndexFeatureFile again to optimize access to PASS-filtered VCF files. VariantFiltration further refines SNP variations. The annotated SNP and Indels variations that passed the filter are ready for analysis.</p>
        </div>

        <a href="{{url_for('homepage')}}" class="arrow" id="index">&#10132; Go to HomePage</a>
        <a href="{{url_for('demo')}}" class="arrow" id="demo">Go to Demo &#10132;</a>
    </div>

{% endblock %}